vignettes/articles/glitter_for_Wikidata.Rmd

```diff
@@ -18,178 +18,225 @@ knitr::opts_chunk$set(
 
-The resulting table may then be used easily with (for instance) package `leaflet`:
+We can thus complete our query this way:
 
-```r
-leaflet::leaflet(stations_metro_Lyon) %>%
-  leaflet::addTiles() %>%
-  leaflet::addCircles(popup=~stations_label)
+```{r query_prop_qualifiers}
+stations_adjacency=spq_init() %>% 
+  spq_add("?station wdt:P16 wd:Q1552") %>% 
+  spq_add("?station wdt:P31 wd:Q928830") %>%
+  spq_add("?station wdt:P625 ?coords") %>%
+  spq_add("?station p:P197 ?statement") %>%          # added instruction
+  spq_add("?statement ps:P197 ?adjacent") %>%        # added instruction
+  spq_add("?statement pq:P81 ?line") %>%             # added instruction
+  spq_add("?statement pq:P5051 ?direction") %>%      # added instruction
+  spq_label("station", "adjacent", "line", "direction",.overwrite=TRUE) %>% 
+  spq_select(-statement) %>% 
+  spq_perform() %>% 
+  na.omit() %>% 
+  select(coords,station,adjacent,line,direction)
+
+head(stations_adjacency)
 ```
 
+Now, we would like to put the stations **in the right order** so that we will be able to form the connecting lines. 
 
-# Example 2 : cities around Lyon
+This **data-wrangling part is a bit tricky** though not directly due to any glitter-related operation).
 
-Now, let's imagine that we are interested in the cities in a 200km radius around Lyon.
+We define a function `form_line()` which will put the rows in the table of stations in the correct order.
 
-```{r}
-find_item("Lyon")
-find_item("city")
+```{r form_line}
+form_line=function(adjacencies,direction){
+  N=nrow(adjacencies)
+  num=rep(NA,N)
+  ind=which(adjacencies$adjacent==direction)
+  i=N
+  num[ind]=i
+  while(i>1){
+    indnew=which(adjacencies$adjacent==adjacencies$station[ind])
+    ind=indnew
+    i=i-1
+    num[ind]=i
+  }
+  adjacencies=adjacencies %>% 
+    mutate(num=num) %>%
+    arrange(num) 
+  adjacencies=c(adjacencies$station,
+                direction)
+  return(adjacencies)
+}
 ```
 
-We could start exploring Wikidata with this query which finds all items that are instances ("wdt:P31") of "city" or of any subclass ("wdt:P279") of "city" . This query might return many items so that it seems reasonable to limit the number of items retrieved for now with the argument `limit`
+Now let's **apply this function to all lines and directions possible**:
 
-```{r}
-spq_init() %>% 
-  spq_add("?city wdt:P31/wdt:P279* wd:Q515", .label="?city") %>%
-  spq_head(n = 10) %>% 
-  spq_perform()
+```{r calc_lines}
+stations_lines=stations_adjacency %>% 
+  sf::st_drop_geometry() %>% # make this a regular tibble, not sf
+  group_by(direction,line) %>% 
+  na.omit() %>% 
+  tidyr::nest(.key="adj") %>% # have nested "adj" table for each direction-line
+  mutate(station=purrr::map(.x=adj,.y=direction,
+                            ~form_line(.x,.y))) %>% 
+  tidyr::unnest(cols="station") %>% 
+  ungroup() %>% 
+  left_join(unique(stations_coords), # get corresponding coordinates
+            by=c("station")) %>%
+  na.omit()
```

Comment:
comment on the usage of this function?

Reply:
Something like "We use left_join() to complete the table ordering the stations into lines with the coordinates of stations."